pkg/common/checker.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"reflect"
+
+	"github.com/kr/pretty"
+	"github.com/sergi/go-diff/diffmatchpatch"
+	"gopkg.in/check.v1"
+)
+
+type diffChecker struct {
+	*check.CheckerInfo
+}
+
+// DeepEquals is a GoCheck checker that does a diff between two objects and
+// pretty-prints any difference between the two. It can act as a substitute
+// for DeepEquals.
+var DeepEquals check.Checker = &diffChecker{
+	&check.CheckerInfo{Name: "Diff", Params: []string{"obtained", "expected"}},
+}
+
+// Check performs a diff between two objects provided as parameters, and
+// returns either true if the objects are identical, or false otherwise. If
+// it returns false, it also uses 'pretty' to print the differences between
+// the objects both with a highlighted list of differences and with diff
+// formatted output where unexpected output is highlighted in green, and
+// output that is expected but not observed is highlighted in red.
+func (checker *diffChecker) Check(params []interface{}, names []string) (result bool, error string) {
+	if len(params) != 2 {
+		return false, "Not comparing two elements!"
+	}
+
+	if reflect.DeepEqual(params[0], params[1]) {
+		return true, ""
+	}
+
+	diff := pretty.Diff(params[0], params[1])
+	if len(diff) == 0 {
+		return false, "Couldn't pretty-print difference"
+	}
+
+	pretty.Println("----------------------------------------------------------------------")
```

Comment:
Could we pack all checker output to string and return it with result instead of printing it from within the function? No existing checkers seem to do so: https://github.com/go-check/check/blob/v1/checkers.go and it would be nice to follow the convention.

Reply:
Thanks for the feedback, I tidied this up and will push soon.